netutil/netutil.go

```diff
@@ -54,3 +57,164 @@ func privateNetworkInterfaces(all []net.Interface, fallback []string, logger log
 	}
 	return privInts
 }
+
+// GetFirstAddressOf returns the first IPv4/IPV6 address of the supplied interface names, omitting any link-local addresses.
+func GetFirstAddressOf(names []string, logger log.Logger, enableInet6 bool) (string, error) {
+	return getFirstAddressOf(names, logger, getInterfaceAddresses, enableInet6)
+}
+
+// NetworkInterfaceAddressGetter matches the signature of net.InterfaceByName() to allow for test mocks.
+type NetworkInterfaceAddressGetter func(name string) ([]netip.Addr, error)
+
+// getFirstAddressOf returns the first IPv4/IPV6 address of the supplied interface names, omitting any link-local addresses.
+func getFirstAddressOf(names []string, logger log.Logger, interfaceAddrsFunc NetworkInterfaceAddressGetter, enableInet6 bool) (string, error) {
+	var ipAddr netip.Addr
+
+	// When passing an empty list of interface names, we select all interfaces.
+	if len(names) == 0 {
+		infs, err := net.Interfaces()
+		if err != nil {
+			return "", fmt.Errorf("failed to get interface list and no interface names supplied: %w", err)
+		}
+		names = make([]string, len(infs))
+		for i, v := range infs {
+			names[i] = v.Name
+		}
+	}
+
+	level.Debug(logger).Log("msg", "looking for addresses", "inf", fmt.Sprintf("%s", names), "inet6enabled", enableInet6)
+
+	for _, name := range names {
+		addrs, err := interfaceAddrsFunc(name)
+		if err != nil {
+			level.Warn(logger).Log("msg", "error getting addresses for interface", "inf", name, "err", err)
+			continue
+		}
+
+		if len(addrs) <= 0 {
+			level.Warn(logger).Log("msg", "no addresses found for interface", "inf", name, "err", err)
+			continue
+		}
+		if ip := filterBestIP(addrs, enableInet6); ip.IsValid() {
+			// Select the best between what we've received
+			ipAddr = filterBestIP([]netip.Addr{ip, ipAddr}, enableInet6)
+		}
+		level.Debug(logger).Log("msg", "detected highest quality address", "ipAddr", ipAddr.String(), "inf", name)
+		if ipAddr.IsLinkLocalUnicast() || !ipAddr.IsValid() {
+			level.Debug(logger).Log("msg", "ignoring address", "ipAddr", ipAddr.String(), "inf", name)
+			continue
+		}
+
+		if enableInet6 {
+			if ipAddr.Is6() {
+				return ipAddr.String(), nil
```

Comment:
When we return an ipv6 addr, we would need to add `[]` around them, to avoid errors like this
```
level=warn ts=2022-12-20T23:20:40.248988657Z caller=logging.go:86 orgID=application msg="POST /loki/api/v1/push (500) 347.86µs Response: \"rpc error: code = Unavailable desc = connection error: desc = \\\"transport: Error while dialing dial tcp: address fd01:0:0:4::29:9095: too many colons in address\\\"\\n\" 
```

Reply:
Good catch.  So the `[]` is not the slice, but just noting the inet6 address, correct?

I think `return fmt.Sprintf("[%s]", ipAddr.String()), nil` would suffice.  Thoughts from the group?

Reply:
That should be fine. The RFC of IPv6 says the string representation should be without brackets when you just display the ip, but in case you represent a "socket" (IP+port) you must add brackets around your IP then add the port [xxxx::yyyy]:3100. So that's also the case for a URI like http://[xxxx::yyyy]:3100

Reply:
@zalegrala Any chance to get this change and the PR merged soon?

Reply:
The change seems simple enough, but I hesitate slightly.  Since we're returning a string that used to be just an address, and now we want to conditionally return a formatted address based on how we "know" the caller will be using this, seems a little strange to me.

Just wondering if returning the formatted string when IPv6 is present it the right change to make.

Reply:
Official RFC about IPv6 + port string representation https://www.rfc-editor.org/rfc/rfc5952#section-6

That could help to take a decision. To me the function return an IP, but in that case the function in charge to concatenate the port should be conditioned by IP family. If not possible, adding bracket for IPv6 all the time seems acceptable.

Reply:
I think I'd go with returning the address without brackets and relying on the user to properly format host+port, for example using [`net.JoinHostPort()`](https://pkg.go.dev/net#JoinHostPort).

Reply:
I tend to agree @xperimental.  Conveniently, `net.JoinHostPort()` operates on strings.

Given that, the various projects that want to test this implementation will need to make updates in order to test this out.   I've pushed `zalegrala/tempo:inet6-dirty` but haven't had a chance to test it.

Reply:
I can test for you on tempo by removing the address field I added ;)

Reply:
Fyi my next candidate is loki

Reply:
Hum... Unfortunately that seems not working.

```
level=error ts=2023-02-07T10:26:05.057010236Z caller=rate_limited_logger.go:27 msg="pusher failed to consume trace data" err="rpc error: code = Unavailable desc = connect
ion error: desc = \"transport: Error while dialing dial tcp: address 2600:1f10:4f41:9d02:210c::f:9095: too many colons in address\""
level=warn ts=2023-02-07T10:26:05.217962589Z caller=tcp_transport.go:438 component="memberlist TCPTransport" msg="WriteTo failed" addr=[2600:1f10:4f41:9d02:40b8::4]:7946
err="dial tcp [2600:1f10:4f41:9d02:40b8::4]:7946: i/o timeout"
level=warn ts=2023-02-07T10:26:05.224162591Z caller=pool.go:184 msg="removing distributor_pool failing healthcheck" addr=2600:1f10:4f41:9d02:210c::e:9095 reason="rpc erro
r: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp: address 2600:1f10:4f41:9d02:210c::e:9095: too many colons in address\""
level=warn ts=2023-02-07T10:26:05.22432428Z caller=pool.go:184 msg="removing distributor_pool failing healthcheck" addr=2600:1f10:4f41:9d02:210c::f:9095 reason="rpc error
: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp: address 2600:1f10:4f41:9d02:210c::f:9095: too many colons in address\""
level=error ts=2023-02-07T10:26:05.925358062Z caller=rate_limited_logger.go:27 msg="pusher failed to consume trace data" err="rpc error: code = Unavailable desc = connect
ion error: desc = \"transport: Error while dialing dial tcp: address 2600:1f10:4f41:9d02:210c::f:9095: too many colons in address\""
level=error ts=2023-02-07T10:26:06.488881128Z caller=rate_limited_logger.go:27 msg="pusher failed to consume trace data" err="rpc error: code = Unavailable desc = connect
ion error: desc = \"transport: Error while dialing dial tcp: address 2600:1f10:4f41:9d02:210c::f:9095: too many colons in address\""
```

I replaced `prefer_inet6` by `enable_inet6` and removed :
```yaml
          ingester:
            lifecycler:
               address: tempodb-ingester-discovery
```

Can you let me know if you update your docker image ? I had to adapt the config and remove deprecated fields like `search_enabled`...

Reply:
This is a very hacky fix that I had used to get it to work in loki. Since the i.lifecycler.Addr already has the address and the port concatenated - https://github.com/shwetaap/loki/blame/ipv6dskitoperator/pkg/ingester/ingester.go#L291-L294
I am not sure how the user needs to fix this, shouldnt it be taken care of in the code automatically by adding the [] depending on if the address used is ipv6?

Reply:
Apologies, I rushed the image out.  On the tempo side, I left a more detailed message in https://github.com/grafana/tempo/issues/1544